Source/Core/VideoCommon/TextureConversionShader.cpp

```diff
@@ -720,4 +724,526 @@ const char* GenerateEncodingShader(u32 format, APIType ApiType)
+        uint val = texelFetch(s_input_buffer, int(buffer_pos)).x;
+        uint i;
+        if ((coords.x & 1u) == 0u)
+          i = Convert4To8((val >> 4));
+        else
+          i = Convert4To8((val & 0x0Fu));
+
+        uvec4 color = uvec4(i, i, i, i);
+        vec4 norm_color = vec4(color) / 255.0;
+
+        imageStore(output_image, ivec3(ivec2(coords), 0), norm_color);
+      }
+
+      )"},
+     {GX_TF_IA4, BUFFER_FORMAT_R8_UINT, 0, 8, 8, false,
+      R"(
+      layout(local_size_x = 8, local_size_y = 8) in;
+
+      void main()
+      {
+        uvec2 coords = gl_GlobalInvocationID.xy;
+
+        // Tiled in 8x4 blocks, 8 bits per pixel
+        uint buffer_pos = GetTiledTexelOffset(uvec2(8u, 4u), coords);
+        uint val = texelFetch(s_input_buffer, int(buffer_pos)).x;
+        uint i = Convert4To8((val & 0x0Fu));
+        uint a = Convert4To8((val >> 4));
+        uvec4 color = uvec4(i, i, i, a);
+        vec4 norm_color = vec4(color) / 255.0;
+
+        imageStore(output_image, ivec3(ivec2(coords), 0), norm_color);
+      }
+      )"},
+     {GX_TF_I8, BUFFER_FORMAT_R8_UINT, 0, 8, 8, false,
+      R"(
+      layout(local_size_x = 8, local_size_y = 8) in;
+
+      void main()
+      {
+        uvec2 coords = gl_GlobalInvocationID.xy;
+
+        // Tiled in 8x4 blocks, 8 bits per pixel
+        uint buffer_pos = GetTiledTexelOffset(uvec2(8u, 4u), coords);
+        uint i = texelFetch(s_input_buffer, int(buffer_pos)).x;
+        uvec4 color = uvec4(i, i, i, i);
+        vec4 norm_color = vec4(color) / 255.0;
+
+        imageStore(output_image, ivec3(ivec2(coords), 0), norm_color);
+      }
+      )"},
+     {GX_TF_IA8, BUFFER_FORMAT_R16_UINT, 0, 8, 8, false,
+      R"(
+      layout(local_size_x = 8, local_size_y = 8) in;
+
+      void main()
+      {
+        uvec2 coords = gl_GlobalInvocationID.xy;
+
+        // Tiled in 4x4 blocks, 16 bits per pixel
+        uint buffer_pos = GetTiledTexelOffset(uvec2(4u, 4u), coords);
+        uint val = texelFetch(s_input_buffer, int(buffer_pos)).x;
+        uint a = (val & 0xFFu);
+        uint i = (val >> 8);
+        uvec4 color = uvec4(i, i, i, a);
+        vec4 norm_color = vec4(color) / 255.0;
+        imageStore(output_image, ivec3(ivec2(coords), 0), norm_color);
+      }
+      )"},
+     {GX_TF_RGB565, BUFFER_FORMAT_R16_UINT, 0, 8, 8, false,
+      R"(
+      layout(local_size_x = 8, local_size_y = 8) in;
+
+      void main()
+      {
+        uvec2 coords = gl_GlobalInvocationID.xy;
+
+        // Tiled in 4x4 blocks
+        uint buffer_pos = GetTiledTexelOffset(uvec2(4u, 4u), coords);
+        uint val = texelFetch(s_input_buffer, int(buffer_pos)).x;
+
+        // Swap BE to LE
+        val = ((val >> 8) | (val << 8)) & 0xFFFFu;
+
+        uvec4 color;
+        color.x = Convert5To8(bitfieldExtract(val, 11, 5));
+        color.y = Convert6To8(bitfieldExtract(val, 5, 6));
+        color.z = Convert5To8(val & 0x1Fu);
+        color.a = 255u;
```

Comment:
What do you think about a 565 decoder? It could be shared with the palette decoder.